### PR TITLE
*: implements health check

### PR DESCRIPTION
**Pull request by BusyJay**
This PR provides simple health check implementations referring to the go
version. It provides a `HealthService` to maintain the service statuses
and wake up watchers.

I use a standalone crate to avoid introduce grpc specific code into the grpcio
crate. And I don't reuse grpcio-proto to avoid the dependency of protobuf-build
which just make things complicated by introducing a lot of dependencies and
resulting in a dependency circle.

### CODE REVIEW

**Comment by BusyJay** [comment]
Most code are generated by protobuf and prost, you can just skip them as long as it compiles.

File: health/src/service.rs [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+        StatusCast {
+            state: AtomicUsize::new(VERSION_STEP | (status as usize)),
+            subscribers: Mutex::default(),
+        }
+    }
+
+    /// Updates the status to specified one and update version.
+    fn broadcast(&self, status: ServingStatus) {
+        let mut subscribers = self.subscribers.lock().unwrap();
+        let state = self.state.load(Ordering::Acquire);
```
Comment (sticnarf) [inline health/src/service.rs:42]: Guarded by the `subscribers` mutex, I think it can be `Relaxed`.

File: health/src/service.rs [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        let cur_state = s.cast.state.load(Ordering::Relaxed);
+        if cur_state != s.last_state {
+            let status = state_to_status(cur_state);
+            s.last_state = cur_state;
+            return Poll::Ready(Some(status));
+        }
+
+        let mut subscribers = s.cast.subscribers.lock().unwrap();
+
+        let cur_state = s.cast.state.load(Ordering::Acquire);
```
Comment (sticnarf) [inline health/src/service.rs:83]: It can be `Relaxed` because a mutex has memory fence itself.

File: health/src/service.rs [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+            subscribers: Mutex::default(),
+        }
+    }
+
+    /// Updates the status to specified one and update version.
+    fn broadcast(&self, status: ServingStatus) {
+        let mut subscribers = self.subscribers.lock().unwrap();
+        let state = self.state.load(Ordering::Acquire);
+        let new_state = ((state + VERSION_STEP) & !STATUS_MASK) | (status as usize);
+        self.state.store(new_state, Ordering::Release);
```
Comment (sticnarf) [inline health/src/service.rs:44]: And because L74 loads with `Relaxed`, this can be `Relaxed` as well.

**Comment by BusyJay** [comment]
I add back protobuf-build to make it work well with TiKV. TiKV seems to use custom protobuf patch which doesn't work well with latest published one.

**Review by sticnarf — DISMISSED** [review]
LGTM

**Review by Connor1996 — COMMENTED** [review]
rest LGTM

File: health/src/service.rs [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+/// is any updates.
+struct StatusCast {
+    state: AtomicUsize,
+    subscribers: Mutex<HashMap<u64, Waker>>,
+}
+
+impl StatusCast {
+    fn new(status: ServingStatus) -> StatusCast {
+        StatusCast {
+            state: AtomicUsize::new(VERSION_STEP | (status as usize)),
```
Comment (Connor1996) [inline health/src/service.rs:34]: why we can't use status directly?

File: health/src/service.rs [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+        let cur_state = s.cast.state.load(Ordering::Relaxed);
+        if cur_state != s.last_state {
+            let status = state_to_status(cur_state);
+            s.last_state = cur_state;
+            return Poll::Ready(Some(status));
+        }
+
+        let mut subscribers = s.cast.subscribers.lock().unwrap();
+
+        let cur_state = s.cast.state.load(Ordering::Relaxed);
```
[the remark this answers is not in the document]
Comment (BusyJay) [reply health/src/service.rs:83]: The first check is a fast path. Only section protected by lock can guarantee `state` is not updated by others.

File: health/src/service.rs (hunk 4 above)
Comment (BusyJay) [reply health/src/service.rs:34]: Then you can't solve the ABA problem.

**Review by Connor1996 — DISMISSED** [review]
LGTM

**Pull request merged**